### ISSUE TITLE
Allow upload of dense OTU table

### ISSUE THREAD

**Issue by tobiasgf** [gbif/metabarcoding-data-toolkit-ui#136]
Maybe allow OTU_table to be tall/dense, if this is easy to implement.
If sample total read counts are currently calculated before "melting" broad/sparse tables, then this may be a problem.

**Comment by thomasstjerne** [comment]
Is this in TSV format ?
Or is this maybe covered by the new possibility to upload the OTU table in BIOM format?

**Comment by tobiasgf** [comment]
If we can have it as a TSV then we serve a broader community.
Simply the three fields for all non-zero read-count records (although this may be crystal clear):
`sample_id, OTU_id, read_count`

**Comment by thomasstjerne** [comment]
I am wondering whether a sparse table will actually contain the sample_ids and OTU_ids ?
At least in BIOM, the sparse table contains the indices rather than the actual IDs. It probably depends on the tools they use to generate it.
I think we should get some example data from the people who have asked for this. If they have index-based sparse tables, it may be tedious for them to convert to id-based and vice versa

**Comment by tobiasgf** [comment]
(You meanJust to make sure that I am clear.

### Input (sparse/broad OTU Table)

| OTU_id | Sample1 | Sample2 | Sample3 |
|--------|---------|---------|---------|
| OTU_1  | 10      | 0       | 5       |
| OTU_2  | 0       | 15      | 0       |
| OTU_3  | 8       | 3       | 0       |

### Expected Output (dense/tall table)
After running some transformation code, the output data frame (dense) would be:

| OTU_id | Sample_id | readcount |
|--------|-----------|-----------|
| OTU_1  | Sample1   | 10        |
| OTU_1  | Sample3   | 5         |
| OTU_2  | Sample2   | 15        |
| OTU_3  | Sample1   | 8         |
| OTU_3  | Sample2   | 3         |

This can be done with a few lines of r (or whatever you are using), and will reduce the size of txt files (as dense format excludes the zeros). I will reach out to the user to see if this is also what he was thinking, stay tuned.

**Comment by thomasstjerne** [comment]
No. I mean this. Is it the actual Sample_id and OTU_id that is in the sparse table, or is it the INDEX of the sample or OTU in the associated Sample file or OTU File

### Index-based sparse OTU Table (Both OTU/taxonomy table and Sample table needs to be ordered)

(GitHub doesn´t allow tables without headers....)
| | | |
|--------|-----------|-----------|
| 0  | 0   | 10        |
| 0  | 2   | 5         |
| 1  | 1   | 15        |
| 2  | 0   | 8         |
| 2  | 1   | 3         |

### ID-based sparse OTU Table (No ordering necessary for OTU/taxonomy table and Sample table)

| OTU_id | Sample_id | readcount |
|--------|-----------|-----------|
| OTU_1  | Sample1   | 10        |
| OTU_1  | Sample3   | 5         |
| OTU_2  | Sample2   | 15        |
| OTU_3  | Sample1   | 8         |
| OTU_3  | Sample2   | 3         |

**Comment by tobiasgf** [comment]
Yes, that is clear (although I believe we use sparse/dense differently).

If simply using some custom scripting to go from sparse to dense (broad to tall) from a simple set of text files, I would be cautious not having headers (and/or id's), especially as the sorting would not necessarily be identical between otu-table and sample and taxonomy tables before the transformation is carried out.

I would expect the ID-based (tall/dense) version as something users with really big datasets may choose to upload instead of a huge matrix with 99.9% zeros, and that it would be something based on "ad.hoc." scripting. I am not whether this tall/broad conversion is implemented in any systems or pipelines - i.e. if we are to "expect" anything predefined – apart from BIOM files.

But stay tuned, I will get hold of a real dataset.
